raulin0/backup-juggler
# feat: adds the command-line interface

# Description
This pull request includes several commits that introduce new functionality, update documentation, improve the project's configuration, and add tests for the Backup Juggler project.

# Changes Made
## Commits:
- **bump: version 0.2.0 → 0.3.0**: Bumps the version of the project from 0.2.0 to 0.3.0.
- **docs: updates MkDocs configuration**: Updates the MkDocs configuration file to enhance the Markdown capabilities of the documentation site. It adds markdown extensions for attr_list, admonition, pymdownx.tasklist, and pymdownx.superfences. Additionally, it includes an extra CSS file for custom styles and adds a custom run command in the extra section.
- **docs: update index with backup juggler quick use information**: Adds comprehensive and detailed documentation for the Backup Juggler command-line tool. The documentation covers its purpose, features, and usage instructions. It includes explanations and examples for the do-backups subcommand and the get-size subcommand. It also improves the structure and organization of the documentation.
- **docs: adds initial version of tutorial, installation instructions, and styling**: Combines changes made to tutorial.md, installation.md, and extra.css into a single commit. It updates the tutorial with multiple sources and get-size sections, adds installation instructions, and enhances the CSS styles for better documentation layout.
- **build: updates project dependencies and add script entry**: Updates the project dependencies, adding the toml library and the typer library with the all extras. It also adds a new script entry to the configuration for the bj script.
- **feat: adds Backup Juggler CLI functionality**: Adds the functionality of the Backup Juggler command-line interface (CLI). It includes the do_backups command and the get_size command. The CLI supports a --version option to display the version and a --help option to show the help message. The CLI is implemented using the Typer framework and the output is styled using the rich library.
- **test: adds additional tests for Backup Juggler CLI**: Adds new test cases to ensure the functionality of the Backup Juggler CLI. The tests cover running the 'bj' script successfully, printing version and help messages, performing backups, and calculating the size of sources.
- **test: adds fixture to run Backup Juggler CLI tests**: Adds the 'run_backup_juggler' fixture to facilitate testing of the Backup Juggler CLI. It allows specifying options, subcommands, sources, and destinations as parameters and invokes the CLI command for testing purposes.

# Motivation and Context
These changes improve the functionality, usability, and maintainability of the Backup Juggler project. The added functionality of the CLI enhances the command-line experience for users, allowing them to perform backups and calculate sizes conveniently. The updated documentation provides comprehensive information and examples to guide users in using the tool effectively. The improved project configuration and dependencies ensure a smooth development experience. The additional tests increase the test coverage and help maintain the reliability of the Backup Juggler CLI.

# How to Test
1. Ensure that the project dependencies are installed.
2. Run the test suite to verify the existing functionality and the newly added tests for the Backup Juggler CLI.
3. Refer to the updated documentation for instructions and examples on using the Backup Juggler tool, including the CLI commands.
4. Test the Backup Juggler CLI by running the 'bj' script with different options, subcommands, and sources to perform backups and calculate sizes.
5. Verify that the version and help messages are displayed correctly.
6. Ensure that the Backup Juggler CLI functionalities work as expected and produce the desired results.

# Related Issue(s)
N/A

# Checklist

Please review and merge this pull request at your earliest convenience. Your feedback and suggestions for improvement are highly appreciated. Thank you!
